pkg/skaffold/build/model.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+)
+
+// buildNode models the artifact dependency graph using a set of channels.
+// Each artifact has a status struct that has success and a failure channel which it closes once it completes building by calling either markSuccess or markFailure respectively.
+// This notifies all listeners waiting for this artifact of a successful or failed build.
+// Additionally it has a reference to the channels for each of its dependencies.
+// Calling `waitForDependencies` ensures that all required artifacts' channels have already been closed and as such have finished building before the current artifact build starts.
+type buildNode struct {
+	imageName    string
+	success      chan interface{}
+	failure      chan interface{}
+	dependencies []buildNode
+}
+
+// markSuccess broadcasts a successful build
+func (a *buildNode) markSuccess() {
+	// closing channel notifies all listeners waiting for this build that it succeeded
+	close(a.success)
+}
+
+// markFailure broadcasts a failed build
+func (a *buildNode) markFailure() {
+	// closing channel notifies all listeners waiting for this build that it failed
+	close(a.failure)
+}
+
+// waitForDependencies returns an error if any dependency build fails
+func (a *buildNode) waitForDependencies(ctx context.Context) error {
+	for _, dep := range a.dependencies {
+		// wait for required builds to complete
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-dep.failure:
+			return fmt.Errorf("failed to build required artifact: %q", dep.imageName)
```

Comment:
how do the errors look when this happens? will this surface the actual error encountered from the build?

Reply:
The existing code returns the error from the first element in the input `[]*latest.Artifact` that fails which necessarily isn't in build order. So you might get `couldn't build "artifact1": failed to build required artifact: "artifact2"` as the only error without getting the error for `artifact2` failure.

I have another PR in the works which will change this behavior to instead return all build failures so then you'd get:
```
couldn't build "artifact2": some error message
couldn't build "artifact1": failed to build required artifact: "artifact2"
```